### PR TITLE
[GiHub Usage] Add Gallery UI

### PR DESCRIPTION
**Pull request by mogah**
This PR adds only the first part of the UI. The second part would include Google Analytics.
image

https://github.com/NuGet/NuGetGallery/issues/7203

### CODE REVIEW

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 1]
```diff
@@ -517,6 +517,49 @@
...
+                        <p>
+                            About @(Model.GitHubDependenciesInformation.TotalRepos.ToKiloFormat()) popular GitHub repositories depend on
+                            @(Model.Id), including the following:
+                        </p>
+
+                        @foreach (var dep in Model.GitHubDependenciesInformation.Repos)
+                        {
+                            <div class="row top-buffer">
+                                <div class="col-xs-12">
+                                    <button class="btn btn-gh-repo text-left btn-block"
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml]: I would use `<a href="@dep.Url"` instead--needing to have Javascript in the `onclick` is a little ugly.

Also, make sure this has a `title` property so that screen readers have a description of where this links to.

File: tests/NuGetGallery.Core.Facts/GitHub/GitHubUsageConfigurationFacts.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace NuGetGallery.GitHub
+{
+    public class GitHubUsageConfigurationFacts
+    {
+        private static RepositoryInformation GenRepo(string id, int stars = 100, params string[] nugetDependencies)
```
Comment (scottbommarito) [inline tests/NuGetGallery.Core.Facts/GitHub/GitHubUsageConfigurationFacts.cs]: I would suggest `CreateRepo` instead of `GenRepo`--I haven't seen that abbreviation in our codebase before and less acronyms is less confusion.

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGetGallery.GitHub
```
Comment (scottbommarito) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: Typically what we do is put all of the files in this project into the same `namespace`.
I don't really have a problem with changing this (in fact I think @ryuyu was thinking of doing the same thing) but we should agree, as a team, that this is the right approach.

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 4]
```diff
@@ -504,11 +504,17 @@ public async Task GivenAValidatingPackageThatTheCurrentUserDoesNotOwnThenHideIt(
                 var deprecationService = new Mock<IPackageDeprecationService>();
                 var httpContext = new Mock<HttpContextBase>();
                 var httpCachePolicy = new Mock<HttpCachePolicyBase>();
+                var contentObjectService = new Mock<IContentObjectService>();
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Do you have any tests that check that the repository information is properly passed on to the model if the feature flag is enabled?

File: tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs [hunk 5]
```diff
@@ -29,5 +29,38 @@ public void FormatsUsingExpectedUnit(long bytes, string expected)
...
+            [InlineData(9999, "9.99K")]
+            [InlineData(10_000, "10.0K")]
+            [InlineData(10_990, "10.9K")]
+            [InlineData(99_990, "99.9K")]
+            [InlineData(100_000, "100K")]
+            [InlineData(100_990, "100K")]
+            [InlineData(100_999, "100K")]
+            [InlineData(999_000, "999K")]
+            [InlineData(999_999, "999K")]
+            [InlineData(1_000_000, "1.00M")]
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: Is `1_000_000` a way to write `1000000` in C#?
Never seen that before, looks pretty cool.

File: src/NuGetGallery.Core/GitHub/IGitHubUsageConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        /// <summary>
+        /// Returns a NuGetPackageGitHubInformation object that contains the information about a NuGet package.
+        /// NOTE: If a packageId has no information, the NuGetPackageGitHubInformation's TotalRepos will be 0
+        /// and the Repos list will be empty
+        /// 
+        /// throws an ArgumentException if the packageId is null
+        /// </summary>
+        /// <param name="packageId">NuGet package id, cannot be null</param>
+        /// <returns></returns>
```
Comment (scottbommarito) [inline src/NuGetGallery.Core/GitHub/IGitHubUsageConfiguration.cs]: Either add some information to this tag or remove it.

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 7]
```diff
@@ -752,6 +753,9 @@ public virtual async Task<ActionResult> DisplayPackage(string id, string version
             model.IsCertificatesUIEnabled = _contentObjectService.CertificatesConfiguration?.IsUIEnabledForUser(currentUser) ?? false;
             model.IsAtomFeedEnabled = _featureFlagService.IsPackagesAtomFeedEnabled();
             model.IsPackageDeprecationEnabled = _featureFlagService.IsManageDeprecationEnabled(currentUser);
+            model.IsGitHubUsageEnabled = _featureFlagService.IsGitHubUsageEnabled(currentUser);
+
+            model.GitHubDependenciesInformation = _contentObjectService.GitHubUsageConfiguration.GetPackageInformation(id);
```
Comment (scottbommarito) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Have you considered not putting this data in the model if the feature flag is disabled? I think this is fine but just a thought.

**Comment by joelverhagen** [comment]
Screenshot please

File: src/NuGetGallery/Extensions/NumberExtensions.cs [hunk 8]
```diff
@@ -66,5 +67,40 @@ public static string ToUserFriendlyBytesLabel(this long bytes)
 
             return "0 bytes";
         }
+
+        /// <summary>
+        /// Format the number to a 3 digit representation plus a letter to represent the scale (K for kilo, M for mega, or G for giga)
+        /// </summary>
+        /// <param name="number"></param>
```
Comment (joelverhagen) [inline src/NuGetGallery/Extensions/NumberExtensions.cs]: Fill out param and returns or remove

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 7 above)
Comment (joelverhagen) [reply src/NuGetGallery/Controllers/PackagesController.cs]: I think it's best to make the feature flag more powerful in this regard. 

---
In reply to: [292238816](https://github.com/NuGet/NuGetGallery/pull/7257#discussion_r292238816) [](ancestors = 292238816)

File: src/NuGetGallery/Services/IFeatureFlagService.cs [hunk 9]
```diff
@@ -52,5 +52,13 @@ public interface IFeatureFlagService
         /// Whether the user is able to access the search side-by-side experiment.
         /// </summary>
         bool IsSearchSideBySideEnabled(User user);
+
+        /// <summary>
+        /// Whether a user can see the "GitHub Usage" section in a package's display page as well as 
+        /// well as the added "GitHub Usage count" in the "Statistics" section
+        /// </summary>
+        /// <param name="user"></param>
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/IFeatureFlagService.cs]: delete/fill out param and returns

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: I think we should be consistent with our current approach (most everything in NuGetGallery namespace) until we design good buckets

---
In reply to: [292237967](https://github.com/NuGet/NuGetGallery/pull/7257#discussion_r292237967) [](ancestors = 292237967)

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: license header

**Comment by joelverhagen** [comment]
> namespace NuGetGallery.GitHub

license header

---
Refers to: src/NuGetGallery.Core/GitHub/IGitHubUsageConfiguration.cs:1 in 229d173. [](commit_id = 229d17330ec98397f82642bcd29744ceadacad8c, deletion_comment = False)

File: src/NuGetGallery.Core/GitHub/RepositoryInformation.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+
+namespace NuGetGallery.GitHub
+{
+    public class RepositoryInformation
+    {
+        public RepositoryInformation(
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/GitHub/RepositoryInformation.cs]: Wait you don't need [JsonConstructor] here?

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+            if( repos == null)
+            {
+                throw new ArgumentNullException(nameof(repos));
+            }
+
+            TotalRepos = totalRepos;
+            Repos = repos
+                .OrderByDescending(x => x.Stars)
+                .ThenBy(x => x.Id)
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: nit: StringComparer.OrdinalIgnoreCase

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+using System.Linq;
+
+namespace NuGetGallery.GitHub
+{
+    public class GitHubUsageConfiguration : IGitHubUsageConfiguration
+    {
+        private readonly Dictionary<string, NuGetPackageGitHubInformation> _nuGetPackagesGitHubDependencies;
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: >Dictionary [](start = 25, length = 10)

nit: IReadOnlyDictionary

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                return value;
+            }
+
+            return NuGetPackageGitHubInformation.Empty;
+        }
+
+        private static Dictionary<string, NuGetPackageGitHubInformation> GetNuGetPackagesDependents(IReadOnlyList<RepositoryInformation> repositories)
+        {
+            var dependentsPerPackage = new Dictionary<string, List<RepositoryInformation>>(StringComparer.InvariantCultureIgnoreCase);
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: >InvariantCultureIgnoreCase [](start = 106, length = 26)

OrdinalIgnoreCase

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+                throw new IndexOutOfRangeException(string.Format("{0} cannot have a negative value!", nameof(totalRepos)));
+            }
+
+            if( repos == null)
+            {
+                throw new ArgumentNullException(nameof(repos));
+            }
+
+            TotalRepos = totalRepos;
+            Repos = repos
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: Nit: maybe do this sorting and list truncation in `GitHubUsageConfiguration` so that all the "business logic" is in one and and this model is just dumb

File: tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs [hunk 16]
```diff
@@ -29,5 +29,38 @@ public void FormatsUsingExpectedUnit(long bytes, string expected)
...
+            [InlineData(1_999_000, "1.99M")]
+            [InlineData(99_990_000, "99.9M")]
+            [InlineData(100_990_000, "100M")]
+            [InlineData(999_990_000, "999M")]
+            [InlineData(1_000_000_000, "1.00G")]
+            [InlineData(1_999_000_000, "1.99G")]
+            public void FormatsUsingExpectedUnit(int number, string expected)
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: I would expect K. M seems relatively common. G seems weird though. i have never seen that for abbreviating quantities. Maybe it's just thinking of financial journals but  I would expect "B" for billion.

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 17]
```diff
@@ -517,6 +517,49 @@
...
+                                            onclick="window.open('@dep.Url');">
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml]: Why not just a link `<a>`?

Project URL, license URL, description URLs etc just do a standard navigation

Also, I would recommend rel=nofollow

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 18]
```diff
@@ -688,11 +700,17 @@ public async Task GivenAnAbsoluteLatestVersionItReturnsTheFirstLatestSemVer2()
                 var packageService = new Mock<IPackageService>();
                 var indexingService = new Mock<IIndexingService>();
                 var deprecationService = new Mock<IPackageDeprecationService>();
+                var contentObjectService = new Mock<IContentObjectService>();
+                contentObjectService
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: perhaps CreateController should do this by default or the app code should be resilient to null

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 19]
```diff
@@ -823,13 +847,20 @@ public async Task GivenAValidPackageWithNoVersionThatTheCurrentUserDoesNotOwnItD
...
                     GetConfigurationService(),
                     packageService: packageService,
                     indexingService: indexingService,
-                    deprecationService: deprecationService);
+                    deprecationService: deprecationService,
+                    contentObjectService: contentObjectService);
                 controller.SetCurrentUser(TestUtility.FakeUser);
 
+
                 var package = new Package()
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: no extra space

**Comment by joelverhagen** [comment]
>             private async Task<ActionResult> GetActionResultForPackageStatusAsync(

Maybe I missed it but where do you actually assert the value from the content object service is the same as the value put into the view model

---
Refers to: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs:496 in 229d173. [](commit_id = 229d17330ec98397f82642bcd29744ceadacad8c, deletion_comment = False)

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 20]
```diff
@@ -517,6 +517,49 @@
...
+                        <span>GitHub Usage</span>
+                    </a>
+                </h2>
+            
+                if (Model.GitHubDependenciesInformation.Repos.Any())
+                {
+                    <div class="github-usage panel-collapse collapse" aria-expanded="true" id="github-usage">
+                        <p>
+                            About @(Model.GitHubDependenciesInformation.TotalRepos.ToKiloFormat()) popular GitHub repositories depend on
+                            @(Model.Id), including the following:
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml]: @karann-msft / @anangaur could you check the phrasing here?

File: src/NuGetGallery/Extensions/NumberExtensions.cs [hunk 21]
```diff
@@ -66,5 +67,40 @@ public static string ToUserFriendlyBytesLabel(this long bytes)
...
+        /// <returns></returns>
+        public static string ToKiloFormat(this int number)
+        {
+            if (number >= 1_000_000_000)
+            {
+                return new StringBuilder((number / 1_000_000_000.0f).ToString("F3")) { [4] = 'G' }.ToString();
```
Comment (joelverhagen) [inline src/NuGetGallery/Extensions/NumberExtensions.cs]: Don't think the string builder adds anything useful here. String interpolation will probably be more readly like:

$"{(number / 1_000_000_000.0f):F3}G"

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs (hunk 3 above)
Comment (agr) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: Is code in Gallery GitHub-specific? Should we keep 'GitHub' out of class names everywhere?

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 22]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGetGallery.GitHub
+{
+    public class GitHubUsageConfiguration : IGitHubUsageConfiguration
```
Comment (agr) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: Is code in Gallery GitHub-specific? Should we keep 'GitHub' out of class names everywhere?

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+            if (_nuGetPackagesGitHubDependencies.TryGetValue(packageId, out var value))
+            {
+                return value;
+            }
+
+            return NuGetPackageGitHubInformation.Empty;
+        }
+
+        private static Dictionary<string, NuGetPackageGitHubInformation> GetNuGetPackagesDependents(IReadOnlyList<RepositoryInformation> repositories)
```
Comment (agr) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: If I understand code correctly this whole function can be rewritten as a single LINQ query. Something along those lines (I didn't actually check that it compiles and works as intended):
```C#
return repositories
    .SelectMany(r => r.Dependencies.Select(d => new {Dependency = d, RepositoryInformation: r}))
    .GroupBy(x => x.Dependency, x => x.RepositoryInformation)
    .ToDictionary(g => g.Key, g => new NuGetPackageGitHubInformation(g.Count, g));
```
(code needs sprinkling with `StringComparer.OrdinalIgnoreCase`).

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public class NuGetPackageGitHubInformation
+    {
+        public const int ReposPerPackage = 10;
+
+        public readonly static NuGetPackageGitHubInformation Empty = new NuGetPackageGitHubInformation(
+                                                            0,
+                                                            new List<RepositoryInformation>());
+
+        public NuGetPackageGitHubInformation(int totalRepos, IReadOnlyList<RepositoryInformation> repos)
```
Comment (agr) [inline src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: `totalRepos` duplicates `repos.Count` here isn't it?

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGetGallery.GitHub
+{
+    public class GitHubUsageConfiguration : IGitHubUsageConfiguration
+    {
+        private readonly Dictionary<string, NuGetPackageGitHubInformation> _nuGetPackagesGitHubDependencies;
+
+        public GitHubUsageConfiguration(IReadOnlyList<RepositoryInformation> repositories)
```
Comment (agr) [inline src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: Please prefer `IReadOnlyCollection` over `IReadOnlyList` everywhere where you don't need access by index.

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+            TotalRepos = totalRepos;
+            Repos = repos
+                .OrderByDescending(x => x.Stars)
+                .ThenBy(x => x.Id)
+                .Take(ReposPerPackage)
+                .ToList();
+        }
+
+        public int TotalRepos { get; }
```
Comment (agr) [inline src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: Same comment as for the argument in constructor. It seems it is not needed or should be implemented as ` => Repos.Count`.

File: src/NuGetGallery.Services/Storage/ContentObjectService.cs [hunk 27]
```diff
@@ -46,9 +49,14 @@ public async Task Refresh()
             TyposquattingConfiguration =
                await Refresh<TyposquattingConfiguration>(ServicesConstants.ContentNames.TyposquattingConfiguration) ??
                new TyposquattingConfiguration();
+
+            var reposCache = await Refresh<IReadOnlyList<RepositoryInformation>>(ServicesConstants.ContentNames.NuGetPackagesGitHubDependencies) ??
+                             Array.Empty<RepositoryInformation>();
```
Comment (agr) [inline src/NuGetGallery.Services/Storage/ContentObjectService.cs]: Layout/indentation is inconsistent with previous statement.

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs (hunk 14 above)
Comment (loic-sharma) [reply src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: Don't we use `ToLowerInvariant` for package IDs for things like the flat container? How do you choose between `OrdinalIgnoreCase` vs `InvariantCultureIgnoreCase`?

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs (hunk 15 above)
Comment (mogah) [reply src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: Loïc and I felt like the `ReposPerPackage` and the filtering would be more appropriate in the class since it's all information that is tightly related to it

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs (hunk 26 above)
Comment (mogah) [reply src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: No this is not, since the Repos collection is a trimmed version of the collection received in the constructor

File: src/NuGetGallery.Core/GitHub/RepositoryInformation.cs (hunk 11 above)
Comment (mogah) [reply src/NuGetGallery.Core/GitHub/RepositoryInformation.cs]: Apparently not 😮

File: src/NuGetGallery/Extensions/NumberExtensions.cs [hunk 28]
```diff
@@ -66,5 +67,40 @@ public static string ToUserFriendlyBytesLabel(this long bytes)
...
         }
+
+        /// <summary>
+        /// Format the number to a 3 digit representation plus a letter to represent the scale (K for kilo, M for mega, or G for giga)
+        /// </summary>
+        /// <param name="number"></param>
+        /// <returns></returns>
+        public static string ToKiloFormat(this int number)
+        {
+            if (number >= 1_000_000_000)
```
Comment (agr) [inline src/NuGetGallery/Extensions/NumberExtensions.cs]: The list of values you compare against can be explicitly declared, might increase the readability:
```C#
var data  = new[]{  // this can be static
    new { Threshold = 1_000_000_000, Transform = n => n / 1_000_000_000f, Format = "{0:G3}G" }, // should it be "B" for billions? What's "G"?
    new { Threshold = 1_000_000, Transform = n => n / 1_000_000f, Format = "{0:G3}M" },
    new { Threshold = 1_000, Transform = n => n / 1_000f, Format = "{0:G3}K" },
    new { Threshold = int.MinValue, Transform = n => n, Format = "{0}" },
};

return data
    .First(d => number >= d.Threshold)
    .Select(d => string.Format(d.Format, d.Transform(number)));
```

File: src/NuGetGallery/Extensions/NumberExtensions.cs [hunk 29]
```diff
@@ -66,5 +67,40 @@ public static string ToUserFriendlyBytesLabel(this long bytes)
...
+            {
+                return new StringBuilder((number / 1_000_000_000.0f).ToString("F3")) { [4] = 'G' }.ToString();
+            }
+
+            if (number >= 100_000_000)
+            {
+                return (number / 1_000_000) + "M";
```
Comment (agr) [inline src/NuGetGallery/Extensions/NumberExtensions.cs]: Integer division uses `floor` to round while float division with subsequent formatting with `F2` uses regular rounding.
I.e. 123_567_890 will produce `123M` while 12_356_789 will produce `12.4M`. That looks inconsistent.

File: src/NuGetGallery/Extensions/NumberExtensions.cs [hunk 30]
```diff
@@ -66,5 +67,40 @@ public static string ToUserFriendlyBytesLabel(this long bytes)
...
+            {
+                return (number / 1_000) + "K";
+            }
+
+            if (number >= 1000)
+            {
+                return new StringBuilder((number / 1_000.0f).ToString(number >= 10_000 ? "F2" : "F3")) { [4] = 'K' }.ToString();
+            }
+
+            return number.ToString("#,0");
```
Comment (agr) [inline src/NuGetGallery/Extensions/NumberExtensions.cs]: This will produce number with thousand separators, all other options don't use them.

File: tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs (hunk 16 above)
Comment (agr) [inline tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: Missing tests for negative numbers.

Comment (agr) [inline tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: I might be nitpicking here, but I have concerns regarding the readability of the output: when you presents a column of numbers with decimal points, you'd want those decimal points aligned:
```
23.4M
1.23M
```
There is an order of magnitude difference between numbers which is not immediately obvious.

Comment (agr) [inline tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: Probably need tests to define/validate rounding rules.

File: tests/NuGetGallery.Core.Facts/GitHub/GitHubUsageConfigurationFacts.cs [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace NuGetGallery
+{
+    public class GitHubUsageConfigurationFacts
+    {
+        private static RepositoryInformation GenRepo(string id, int stars = 100, params string[] nugetDependencies)
```
Comment (agr) [inline tests/NuGetGallery.Core.Facts/GitHub/GitHubUsageConfigurationFacts.cs]: Does it compile in our build pipelines? We don't have the latest compiler there, default argument in the middle of argument list might not be supported.

File: src/NuGetGallery/Extensions/NumberExtensions.cs (hunk 21 above)
Comment (mogah) [reply src/NuGetGallery/Extensions/NumberExtensions.cs]: This is technically an F2, but I use F3 since there are weird rounding problems and then override the last character with the Unit scale

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs (hunk 4 above)
Comment (mogah) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: What do you mean?

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs (hunk 15 above)
Comment (loic-sharma) [reply src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: Yup, especially since consumers of this class expect `Repos` to be ordered.

File: tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs (hunk 5 above)
Comment (mogah) [reply tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: Yeah it's nice. I used to do it in Java, tried it in C#, it worked 😄

File: src/NuGetGallery/Extensions/NumberExtensions.cs (hunk 29 above)
Comment (mogah) [reply src/NuGetGallery/Extensions/NumberExtensions.cs]: 12_356_789 produces 12.3M

Comment (agr) [reply src/NuGetGallery/Extensions/NumberExtensions.cs]: Missed the fact that last digit was overwritten with K/M/G

File: src/NuGetGallery/Extensions/NumberExtensions.cs [hunk 32]
```diff
@@ -75,32 +76,22 @@ public static string ToUserFriendlyBytesLabel(this long bytes)
...
-            }
-
-            if (number >= 100_000)
-            {
-                return (number / 1_000) + "K";
-            }
+                new { Threshold = 1_000_000_000, Transform = new Func<int, float>((x) => x / 1_000_000_000f), Format = "{0:F3}B", Trim = true },
+                new { Threshold = 100_000_000, Transform =  new Func<int, float>((x) => x / 1_000_000), Format = "{0:F0}M", Trim = false},
+                new { Threshold = 10_000_000, Transform =  new Func<int, float>((x) => x / 1_000_000f), Format = "{0:F2}M", Trim = true},
+                new { Threshold = 1_000_000, Transform =  new Func<int, float>((x) => x / 1_000_000f), Format = "{0:F3}M", Trim = true},
```
Comment (agr) [inline src/NuGetGallery/Extensions/NumberExtensions.cs]: Why not `G3` format? This will leave 3 significant digits and will cover 1M/10M/100M cases in a single line.

File: src/NuGetGallery/Extensions/NumberExtensions.cs (hunk 8 above)
Comment (mogah) [reply src/NuGetGallery/Extensions/NumberExtensions.cs]: I was just following the file style. The param is the int the method is applied to since it is an extension

File: src/NuGetGallery.Services/ServicesConstants.cs [hunk 33]
```diff
@@ -39,6 +39,7 @@ public static class ContentNames
             public static readonly string CertificatesConfiguration = "Certificates-Configuration";
             public static readonly string SymbolsConfiguration = "Symbols-Configuration";
             public static readonly string TyposquattingConfiguration = "Typosquatting-Configuration";
+            public static readonly string NuGetPackagesGitHubDependencies = "GitHubUsage.v1";
```
Comment (shishirx34) [inline src/NuGetGallery.Services/ServicesConstants.cs]: why `.v1`, what other iterations are planned? why not name it `GitHubUsage-Configuration` same as all other configs?

File: src/NuGetGallery.Services/Storage/ContentObjectService.cs [hunk 34]
```diff
@@ -46,9 +48,15 @@ public async Task Refresh()
             TyposquattingConfiguration =
                await Refresh<TyposquattingConfiguration>(ServicesConstants.ContentNames.TyposquattingConfiguration) ??
                new TyposquattingConfiguration();
+
+            var reposCache = 
+                await Refresh<IReadOnlyCollection<RepositoryInformation>>(ServicesConstants.ContentNames.NuGetPackagesGitHubDependencies) ??
```
Comment (shishirx34) [inline src/NuGetGallery.Services/Storage/ContentObjectService.cs]: I guess the background job will be writing the repos cache to this location?

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 35]
```diff
@@ -517,6 +517,50 @@
...
 
+            @if(Model.IsGitHubUsageEnabled)
+            {
+                <h2>
+                    <a role="button" data-toggle="collapse" data-target="#github-usage"
+                       aria-expanded="false" aria-controls="github-usage" id="show-github-usage">
+                        <i class="ms-Icon ms-Icon--ChevronRight" aria-hidden="true"></i>
+                        <span>GitHub Usage</span>
+                    </a>
+                </h2>
```
Comment (shishirx34) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml]: I am assuming this section will also be collapsed by default, what do you think of adding an overview details to the header to the right...

Something like

```
> GitHub Usage                                                      1377 Repositories
```
The overview details can be augmented with other details if needed.

It would also be a good idea to add this to other sections(#dependencies of the package for Dependencies section, #versions for Version History section.

@karann-msft, @anangaur - thoughts?

File: src/NuGetGallery/Extensions/NumberExtensions.cs (hunk 8 above)
Comment (joelverhagen) [reply src/NuGetGallery/Extensions/NumberExtensions.cs]: We have been inconsistent/low quality in the past. This is a place where being inconsistent is OK IMO.

File: src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs (hunk 15 above)
Comment (joelverhagen) [reply src/NuGetGallery.Core/GitHub/NuGetPackageGitHubInformation.cs]: Sounds good.

File: src/NuGetGallery.Services/ServicesConstants.cs (hunk 33 above)
Comment (mogah) [reply src/NuGetGallery.Services/ServicesConstants.cs]: @loic-sharma Thought that it might be better to name it this way (same format as downloads.v1.json file) since we might change the format in the futur

File: src/NuGetGallery.Services/Storage/ContentObjectService.cs (hunk 34 above)
Comment (mogah) [reply src/NuGetGallery.Services/Storage/ContentObjectService.cs]: yes

File: src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs (hunk 22 above)
Comment (mogah) [reply src/NuGetGallery.Core/GitHub/GitHubUsageConfiguration.cs]: The whole project is GitHub specific. Indexing other Git repos hosting providers was a non-goal

File: tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs (hunk 16 above)
Comment (mogah) [reply tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: Yeah K is for Kilo, M for Mega, so G is for Giga, but you're the 3rd person to notice this, so I guess B is more used than G ^^

File: src/NuGetGallery.Services/ServicesConstants.cs (hunk 33 above)
Comment (loic-sharma) [reply src/NuGetGallery.Services/ServicesConstants.cs]: This file is closer to `downloads.v1.json` and `owners.v1.json` as it is generated periodically by a background job.

We didn't add a `-Configuration` suffix as this file contains data exported from GitHub (it's not configs).

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml (hunk 35 above)
Comment (loic-sharma) [reply src/NuGetGallery/Views/Packages/DisplayPackage.cshtml]: This is an interesting idea. I think I'd prefer the information to be next to the title:

```
> GitHub Usage (1377 Repositories)
```

We'll ask the UX expert about this suggestion, thanks!

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 36]
```diff
@@ -517,6 +517,50 @@
...
+                                </div>
+                            </div>
+
+                        }
+                    </div>
+                }
+                else
+                {
+                    <p class="github-usage panel-collapse collapse" aria-expanded="true" id="github-usage">
+                        This package is not used by any popular GitHub repository.
```
Comment (loic-sharma) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml]:
```suggestion
                        This package is not used by any popular GitHub repositories.
```

File: tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs (hunk 16 above)
Comment (loic-sharma) [reply tests/NuGetGallery.Facts/Extensions/NumberExtensionsFacts.cs]: @joelverhagen Why are you reading financial journals with billions? 😆
